src/main/java/com/amplitude/Retry.java

```diff
@@ -0,0 +1,253 @@
+        result = new retryEventsOnceResult();
+        result.shouldRetry = shouldRetry;
+        result.shouldReduceEventCount = shouldReduceEventCount;
+        result.eventIndicesToRemove = eventIndicesToRemove;
+        return result;
+    }
+}
+
+class Retry {
+    // Use map to record the events are currently in retry queue.
+    private static Map<String, Map<String, List<Event>>> idToBuffer = new HashMap<String, Map<String, List<Event>>>();
+    private static int eventsInRetry = 0;
+
+    // Helper method to get event list from idToBuffer
+    private static List<Event> getRetryBuffer(String userId, String deviceId) {
+        return (idToBuffer.get(userId) != null) ? idToBuffer.get(userId).get(deviceId) : null;
+    }
+
+    private static List<Event> pruneEvent(List<Event> events) {
+        List<Event> prunedEvents = new ArrayList<>();
+        // If we already have the key value pair for the current event in idToBuffer,
+        // We just add into the events list and deal with it later otherwise we should add it to prunedEvents and return back
+        for (Event event : events) {
+            String userId = event.userId;
+            String deviceId = event.deviceId;
+            if ((userId != null && userId.length() > 0) || (deviceId != null && deviceId.length() > 0)) {
+                List<Event> currentBuffer = getRetryBuffer(userId, deviceId);
+                if (currentBuffer != null) {
+                    currentBuffer.add(event);
+                    eventsInRetry++;
+                } else {
+                    prunedEvents.add(event);
+                }
+            }
+        }
+        return prunedEvents;
+    }
+
+    // Cleans up the id in the buffer map if the job is done
+    private static void cleanUpBuffer(String userId, String deviceId) {
+        Map<String, List<Event>> deviceToBufferMap = idToBuffer.get(userId);
+        if (deviceToBufferMap == null) {
+            return;
+        }
+        List<Event> eventsToRetry = deviceToBufferMap.get(deviceId);
+        if (eventsToRetry != null && eventsToRetry.size() == 0) {
+            deviceToBufferMap.remove(deviceId);
+        }
+        if (deviceToBufferMap.size() == 0) {
+            idToBuffer.remove(userId);
+        }
+    }
+
+    private static int[] JSONArrayToIntArray(JSONArray jsonArray) {
+        int[] arr = new int[jsonArray.length()];
+        for (int i = 0; i < arr.length; ++i) {
+            arr[i] = jsonArray.optInt(i);
+        }
+        return arr;
+    }
+
+    private static int[] collectInvalidEventIndices(Response response) {
+        List<Integer> invalidIndices = new ArrayList<Integer>();
+        if (response.status == Status.INVALID && response.InvalidRequestBody != null) {
+            JSONObject eventsWithInvalidFields = response.InvalidRequestBody.getJSONObject("eventsWithInvalidFields");
+            JSONObject eventsWithMissingFields = response.InvalidRequestBody.getJSONObject("eventsWithMissingFields");
```

Comment:
Seems the logic for both Fields processing is the same, we could extract it to a method and reuse